pkg/services/navtree/navtreeimpl/applinks.go

```diff
@@ -12,13 +14,18 @@ import (
 	"github.com/grafana/grafana/pkg/services/pluginsettings"
 )
 
-func (s *ServiceImpl) getAppLinks(c *models.ReqContext) ([]*navtree.NavLink, error) {
+func (s *ServiceImpl) addAppLinks(treeRoot *navtree.NavTreeRoot, c *models.ReqContext) error {
+	topNavEnabled := s.features.IsEnabled(featuremgmt.FlagTopnav)
 	hasAccess := ac.HasAccess(s.accessControl, c)
 	appLinks := []*navtree.NavLink{}
 
+	if len(s.navigationAppConfig) == 0 {
+		s.readNavigationSettings()
+	}
```

Comment:
Please read/parse settings in `ProvideService` once. This could potentially lead to concurrency issues if called from multiple goroutines at the same time.

Reply:
@marefr how? Is there a init function in services? What's the order? Cfg needs to be loaded before, when is cfg service loaded? Before provide is called?

Reply:
A dependency is supposed to be initialized when your service receive it - so you normally initialize your service in the  `Provide` function.